File: octopoints_flutter/lib/service/impl/RuleServiceImpl.dart
import 'package:octopoints_flutter/db/DbSingleton.dart';
import 'package:octopoints_flutter/db/dao/RuleDao.dart';
import 'package:octopoints_flutter/service/RuleService.dart';
import 'package:octopoints_flutter/service/model/Rule.dart';
import 'package:octopoints_flutter/service/mapper/RuleMapper.dart';

class RuleServiceImpl implements RuleService {
  final RuleDao _ruleDao = DbSingleton().db.ruleDao;

  @override
  Future<Rule> createRule(Rule rule) async {
    int id = await _ruleDao.create(rule.toRuleEntity());
    return rule.setId(id);
  }

  @override
  Future<int> deleteRule(Rule rule) {
    return _ruleDao.remove(rule.toRuleEntity());
  }

  @override
  Future<Rule?> getRuleByMatchId(int id) async {
    return await _ruleDao.getRule(id).then((value) => value.toRuleModel(),);
  }

  @override
  Future<int> updateRule(Rule rule) {
    return _ruleDao.modify([rule.toRuleEntity()]);
  }
}

File: octopoints-kotlin/app/src/main/java/dev/tpavani/octopoints/service/impl/RuleServiceImpl.kt
package dev.tpavani.octopoints.service.impl

import dev.tpavani.octopoints.db.DbSingleton
import dev.tpavani.octopoints.service.RuleService
import dev.tpavani.octopoints.service.model.Rule
import dev.tpavani.octopoints.service.utils.ModelMapper

class RuleServiceImpl : RuleService {
    override suspend fun createRule(rule: Rule): Rule {
        var id: Int = DbSingleton.getDb().ruleDao().create(ModelMapper.toRuleEntity(rule))
        return rule.setId(id)
    }

    override suspend fun deleteRule(rule: Rule) {
        DbSingleton.getDb().ruleDao().remove(ModelMapper.toRuleEntity(rule))
    }

    override suspend fun updateRule(rule: Rule) {
        DbSingleton.getDb().ruleDao().modify(mutableListOf(ModelMapper.toRuleEntity(rule)))
    }

    override suspend fun getRuleByMatchId(id: Int): Rule? {
        return ModelMapper.toRuleModel(DbSingleton.getDb().ruleDao().getRule(id))
    }
}